libraries/chain/controller.cpp

```diff
@@ -1512,7 +1514,14 @@ struct controller_impl {
                // irreversible. Instead, this moves irreversible as much as possible and allows the next maybe_switch_forks call to apply these
                // non-validated blocks. After the maybe_switch_forks call (before next produced block or on next received block), irreversible
                // can then move forward on the then validated blocks.
-               return read_mode == db_read_mode::IRREVERSIBLE || bsp->is_valid();
+               // In irreversible mode, break every ~500ms to allow other tasks (e.g. get_info, SHiP) opportunity to run. There is a post
+               // for every incoming blocks; enough posted tasks to apply all blocks queued to the fork db.
+               if (read_mode == db_read_mode::IRREVERSIBLE) {
+                  if (!replaying && fc::time_point::now() - start > fc::milliseconds(500))
+                     return false;
```

Comment:
When we return false, we will stop processing the second loop below, but I believe that the `fc::raw::pack(*b);` tasks to serialize the blocks for the block log will still be executed, even though the future to get their result will be lost. Is this a performance issue? If it is maybe we could just have a single loop and post the serialization tasks as we apply the blocks?

Reply:
In non-irreversible mode `apply_irreversible_block` does nothing.

Reply:
Thanks, makes sense.
It still makes me uneasy checking code in that possibly exit the loop leaving what could be hundreds of blocks queued for packing, and when we re-enter the loop immediately after we will queue again these hundreds of blocks for packing, and exit after processing a couple blocks (if they are full), and repeat.

If you have a thousand blocks, and process only a couple at a time, many blocks could be packed hundreds of times.

Reply:
I agree. I'm working on a fix.

Reply:
Well, maybe I'm mistaken. We will exit based on time only in irreversible mode, and when in this mode `apply_irreversible_block` does nothing. Still I'm not sure if the rest (`emit`, `blog.append`, `db.commit`, and waiting for the first `pack` tasks) could take enough time to leave a large number of queued `pack` tasks.

Reply:
Also I wonder what happens if we get an exception while processing the 10th block of the branch. We have signaled `irreversible_block`, updated `db` and the blog, but not advanced forkdb's root. 

So would we enter `log_irreversible` and reapply the first 9 blocks which were already applied?

Reply:
I think the two `wlog` in the catch block should be `elog` because they would leave us in an incorrect state.